core/embed/rust/src/protobuf/defs.rs

```diff
@@ -0,0 +1,208 @@
+use core::{mem, slice};
+
+pub struct MsgDef {
+    pub fields: &'static [FieldDef],
+    pub defaults: &'static [u8],
+    pub is_experimental: bool,
+    pub wire_id: Option<u16>,
+    pub offset: u16,
+}
+
+impl MsgDef {
+    pub fn for_name(msg_name: u16) -> Option<Self> {
+        find_msg_offset_by_name(msg_name).map(|msg_offset| unsafe {
+            // SAFETY: We are taking the offset right out of the definitions so we can be
+            // sure it's to be trusted.
+            get_msg(msg_offset)
+        })
+    }
+
+    pub fn for_wire_id(wire_id: u16) -> Option<Self> {
+        find_msg_offset_by_wire(wire_id).map(|msg_offset| unsafe {
+            // SAFETY: We are taking the offset right out of the definitions so we can be
+            // sure it's to be trusted.
+            get_msg(msg_offset)
+        })
+    }
+
+    pub fn field(&self, tag: u8) -> Option<&FieldDef> {
+        self.fields.iter().find(|field| field.tag == tag)
+    }
+}
+
+#[repr(C, packed)]
+pub struct FieldDef {
+    pub tag: u8,
+    flags_and_type: u8,
+    enum_or_msg_offset: u16,
+    pub name: u16,
+}
+
+impl FieldDef {
+    pub fn get_type(&self) -> FieldType {
+        match self.ftype() {
+            0 => FieldType::UVarInt,
+            1 => FieldType::SVarInt,
+            2 => FieldType::Bool,
+            3 => FieldType::Bytes,
+            4 => FieldType::String,
+            5 => FieldType::Enum(unsafe { get_enum(self.enum_or_msg_offset) }),
+            6 => FieldType::Msg(unsafe { get_msg(self.enum_or_msg_offset) }),
+            _ => unreachable!(),
```

Comment:
this is not unreachable though? not sure about style, but istm `panic!` would be more appropriate

Reply:
From the perspective of `defs` this is unreachable without a bug in `pb2py`, and `unreachable!` panics.